### PR TITLE
Cm/update testing

### PR DESCRIPTION
**Pull request by CheezItMan**
## Description
Revised the Intro to testing to remove assertions beyond a note that they are a **thing**.

## Todos
Are any of these items still in progress?
- [x] Splitting the content
- [x] Moving away from die and to... Bill??

## Feedback Requested
Please let me know if you spot a problem.

### CODE REVIEW

File: 00-programming-fundamentals/intro-to-automated-tests.md [hunk 1]
```diff
@@ -124,91 +124,108 @@ We are now setup, but we haven't actually written a test yet.
...
-    # The class of @die should be Die
-    assert @die.class == Die, "There must be a Die class."
+    # the class of @die should be Die
+    @die.class.must_equal Die
   end
 end
 ```
 
-Notice the line `assert @die.class, Die`.  The `assert` method takes two required arguments and an optional message.  If the two required arguments are equal, the test passes, otherwise it fails.  
+Notice the line `@die.class.must_be_instance_of Die`.  Minitest adds the `must_equal` method to all objects and it takes one required argument.  If the required argument is equal to the object `must_equal` is being called on, the expectation passes, otherwise it fails.
```
Comment (droberts-ada) [inline 00-programming-fundamentals/intro-to-automated-tests.md]: Here you say `must_be_instance_of Die`, but above you say `.class.must_equal Die`.

File: 00-programming-fundamentals/intro-to-automated-tests.md [hunk 2]
```diff
@@ -225,93 +242,51 @@ class Die
...
 
 
 #### A Word on Parentheses
 
-Check out the code above.  We are calling assert without using parentheses here:  `assert @die.class == Die, "There must be a Die class."`  Ruby doesn't FORCE you to put parentheses around a method's arguments but [the community-driven style guidelines](https://github.com/bbatsov/ruby-style-guide) suggest that it's good coding style to put parentheses around method arguments **except** for methods part of an internal Domain Specific Language (DSL), or basically the syntax of some kind of framework like... Minitest.  
+Check out the code above.  We are calling `must_equal` without using parentheses here:  `@die.class.must_equal Die`  Ruby doesn't FORCE you to put parentheses around a method's arguments but [the community-driven style guidelines](https://github.com/bbatsov/ruby-style-guide) suggest that it's good coding style to put parentheses around method arguments **except** for methods part of an internal Domain Specific Language (DSL), or basically the syntax of some kind of framework like... Minitest.
```
Comment (droberts-ada) [inline 00-programming-fundamentals/intro-to-automated-tests.md]: Punctuation

**Comment by droberts-ada** [comment]
Also saying `Die` all over the place makes me feel like we're murdering the students. What if someone's grandma just died and this brings back all sorts of bad memories?

File: 00-programming-fundamentals/intro-to-automated-tests.md [hunk 3]
```diff
@@ -124,91 +124,108 @@ We are now setup, but we haven't actually written a test yet.
 
 ### A Tale of Two Styles
 
-In the TDD World there are two styles of testing.  In the first more traditional method people use *assertions* which are statements that check if a value is what it should be.  The other method is a subset of TDD, called Behavior-Driven Development (BDD) which accomplishes the same thing in a more English-friendly fashion that even non-developers can understand. 
+In the TDD World there are two styles of testing.  In the first more traditional method people use *assertions* which are statements that check if a value is what it should be.  The other method is a subset of TDD, called Behavior-Driven Development (BDD) which accomplishes the same thing in a more English-friendly fashion that even non-developers can understand.  At Ada we will use the second BDD style of testing.  You should know assertion-style testing is a thing, and that it accomplishes the same job as our behavior-driven development, but we will not require you to write assertion-style tests.
```
Comment (tildeee) [inline 00-programming-fundamentals/intro-to-automated-tests.md]: I know that BDD was literally written intended to help non-developers understand the code, but is there a way that we can word it so it doesn't use the phrase "that even non-developers can understand"?

Like "business analysts" or something!

I want to transition away from language that reinforces that we're choosing to do something because they are _non-developers_ , because they are/are becoming developers! Also, "technical vs. non-technical" is a weird whole discussion

**Comment by tildeee** [comment]
I agree that it'd be great to move away from the `die` example! A lotta students got weird and morbid with that one

**Comment by CheezItMan** [comment]
@tildeee @droberts-ada 

Ok it's separated out and fixes made.  Any suggestions for replacements for Die, I'd like something simple, but useful with only a few methods.  Dice seemed logical when I wrote this originally.  

I still like it for those reasons, and could simply have them set the seed in testing this time around.

**Comment by CheezItMan** [comment]
Updated and moved away from `die`.

File: 02-intermediate-ruby/intro-to-minitest.md [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+```ruby
+# bill_test.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'bill'
+
+describe "bill" do
+  it "can be created" do
+    bill = Bill.new [{name: 'Coffee', price: 4.75}, {name: 'Pancakes', price: 8.75}]
```
Comment (droberts-ada) [inline 02-intermediate-ruby/intro-to-minitest.md]: It might be simpler / less distracting if it just took an array of prices, rather than messing with names as well.

File: 02-intermediate-ruby/intro-to-minitest.md [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+```
+
+We also need to ensure that if Bill.new is called without an argument it raises an error.  
+
+So we will add another `it` block and use the `must_raise` expectation.
+
+```ruby
+it "will raise an error if created with an invalid argument" do
+  proc {
+    bill = Bill.new
```
Comment (droberts-ada) [inline 02-intermediate-ruby/intro-to-minitest.md]: Rather than "with an invalid argument", could you say something more descriptive? Perhaps "without a list of purchases".

File: 02-intermediate-ruby/intro-to-minitest.md [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+  }.must_raise ArgumentError
+end
+```
+
+**Question**: Look back at the [guidelines for what to test]().  What other kinds of things should we test on the `initialize` method?
```
Comment (droberts-ada) [inline 02-intermediate-ruby/intro-to-minitest.md]: Broken link

File: 02-intermediate-ruby/intro-to-minitest.md [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+Now we need make a method that calculates the bill subtotal.  Write a test (`it` block) that creates an instance of Bill, calls a `subtotal` method and ensure that it returns the correct answer (the sum of the prices).  You can use the `must_equal` matcher.  
+
+#### Check & Verify
+
+Check with your neighbor.  You can find a solution [here.](https://github.com/AdaGold/bill_calculator/commit/12548edbe1d88bce598b642764e7953c10ced204)
+
+### Arrange-Act-Assert
+
+In our tests we will typically follow this pattern:
```
Comment (droberts-ada) [inline 02-intermediate-ruby/intro-to-minitest.md]: Would it make sense to move this section earlier in the resource? It seems a little odd to write a few tests and then talk about what order to do things in a test.

File: 02-intermediate-ruby/intro-to-minitest.md (hunk 5 above)
Comment (droberts-ada) [reply 02-intermediate-ruby/intro-to-minitest.md]: I would also argue that this isn't an interesting behavior to test, since throwing an `ArgumentError` when you forget a parameter is baked into Ruby. There's no line of code you could remove or change to make this test red.

A more interesting scenario might be "what happens when the provided list is empty?"

**Comment by droberts-ada** [comment]
Here is my draft of talking through test cases for Bill: https://gist.github.com/droberts-ada/0edcd9e39196ec7729f8752c879955ee

**Comment by CheezItMan** [comment]
@droberts-ada Updated

File: 00-programming-fundamentals/intro-to-automated-tests.md [hunk 8]
```diff
@@ -67,340 +65,99 @@ true
...
-
-require 'minitest/autorun'
-require 'minitest/reporters'
-require_relative 'die'
-
-class TestDie < MiniTest::Unit::TestCase
-  def test_creation_of_die
-
+class Bill
+  def initialize(item)
```
Comment (kariabancroft) [inline 00-programming-fundamentals/intro-to-automated-tests.md]: should this be `items`? Are we expected an array here?

File: 00-programming-fundamentals/intro-to-automated-tests.md [hunk 9]
```diff
@@ -67,340 +65,99 @@ true
...
+So when we test the Bill class we will need to test both _nominal cases_, where we test that your code works normally, and _edge cases_ where you test the borders of possible inputs.  Anytime you think of a scenario where you say, "Huh that's interesting,"" or "I dunno," you're probably at an _edge case_.
 
-Check again with your partner when you are finished.  You can find a solution [here](https://github.com/AdaGold/die/blob/solution/specs/die_spec.rb)
+#### Nominal Cases
 
+Some nominal cases include:
 
-## Which Style Should I Use?
+- Can I create a `Bill` with a list of normal prices?
+- Given a created bill, does it calculate the subtotal correctly?
```
Comment (kariabancroft) [inline 00-programming-fundamentals/intro-to-automated-tests.md]: Could we either add the comma to each row below or remove from this line?

File: 00-programming-fundamentals/intro-to-automated-tests.md [hunk 10]
```diff
@@ -67,340 +65,99 @@ true
...
+- Are there edge cases to test for with `Bill#tax`, `Bill#total`, or `Bill#subtotal`
 
-[Rake](https://github.com/ruby/rake) is a utility Ruby developers use which lets you set up tasks, like testing and running your programs.  Rake reads the Rakefile and follows the Ruby code in the Rakefile to execute listed tasks.  
+### The Process of Testing: Arrange-Act-Assert
 
-All we are going to use Rake for now is testing, but later Rake will do all sorts of neat stuff for us.
+Testing typically follows this pattern:
+1. Arrange our code with all our variables and inputs:  **Arrange**
+2. Perform an action which we want to test: **Act**
+3. Check with an expectation if it gives the desired result:  **Assert**.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/intro-to-automated-tests.md]: Can you remove the period on this for consistency?

**Comment by CheezItMan** [comment]
Thanks Kari!

File: 02-intermediate-ruby/testing-with-minitest.md [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+  }.must_raise ArgumentError
+end
+```
+
+**Question**: The code above tests for the first situation.  How can you write a tests for the others?  You can find a solution [here](https://github.com/AdaGold/bill_calculator/commit/96f00a74e60d4e4246d0c8a10dfe66dea89cfd6d)
+
+#### How to make it pass?
+
+Now we need to write code to make the tests pass.  Raising an error is new for us right now, so lets look at how we can do so:
+
```
Comment (droberts-ada) [inline 02-intermediate-ruby/testing-with-minitest.md]: I have a few comments about this section:

1. I am nervous about introducing `raise` for the first time here - this is already going to be an overwhelming amount of new stuff for the students, and I think this dilutes our focus
1. You code block above, where you call `Bill.new` with no arguments, is testing built-in ruby functionality. There's not a line of code we could remove or change to make it not raise an argument error, so I would argue it's not something we should be testing.
1. Checking the class of an argument is a classic anti-pattern in dynamically typed languages, and is not a behavior we should be modeling for students. See [this article](https://mikewilliamson.wordpress.com/2013/01/16/duck-typing-in-ruby-thoughts-on-checking-argument-types/) or Metz ch 5.

One way to resolve these issues might be to:

- Re-write the test above to pass in an empty array, and don't add tests for invalid or too few arguments
- Change the `Bill` implementation to be OK with an empty array

The test will pass without us having to add any code, which is not ideal, but I still think it will make this section flow a little smoother.

File: 02-intermediate-ruby/testing-with-minitest.md [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+#### Check & Verify
+
+Check with your neighbor.  You can find a solution [here for the test](https://github.com/AdaGold/bill_calculator/commit/e5bbfe7cd0235e93408e6fae81abf5f2c8624437) [and here for `bill.rb`](https://github.com/AdaGold/bill_calculator/commit/c67fec9271b50fe1d2fd28cf93db343b2dee4f06)
+
+#### Arrange-Act-Assert
+
+Notice our example followed a pattern from our earlier [introduction to automated tests](../00-programming-fundamentals/intro-to-automated-tests.md).  First we created a `Bill`, or **arranged** the situation.  Then by calling the `subtotal` method we **acted**, or performed the action we wanted to test.  Lastly we used the expectation to **assert** that the result of our action was correct.    This is a good example of the _Arrange-Act-Assert_ pattern discussed earlier.  
+
+#### Practice Exercise - Sales Tax
```
Comment (droberts-ada) [inline 02-intermediate-ruby/testing-with-minitest.md]: This feels like a lot of `h4`s (`####`) in a row. Maybe it's fine, but it might be there's an opportunity to add more structure here.

File: 02-intermediate-ruby/testing-with-minitest.md [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+
+Then make the test pass by editing `bill.rb`
+
+#### Check & Verify
+
+Check with your neighbor.  You can find a solution [here for the test](https://github.com/AdaGold/bill_calculator/commit/e5bbfe7cd0235e93408e6fae81abf5f2c8624437) [and here for `bill.rb`](https://github.com/AdaGold/bill_calculator/commit/c67fec9271b50fe1d2fd28cf93db343b2dee4f06)
+
+#### Arrange-Act-Assert
+
+Notice our example followed a pattern from our earlier [introduction to automated tests](../00-programming-fundamentals/intro-to-automated-tests.md).  First we created a `Bill`, or **arranged** the situation.  Then by calling the `subtotal` method we **acted**, or performed the action we wanted to test.  Lastly we used the expectation to **assert** that the result of our action was correct.    This is a good example of the _Arrange-Act-Assert_ pattern discussed earlier.
```
Comment (droberts-ada) [inline 02-intermediate-ruby/testing-with-minitest.md]: I really like this tie-in!

File: 02-intermediate-ruby/testing-with-minitest.md (hunk 11 above)
Comment (CheezItMan) [reply 02-intermediate-ruby/testing-with-minitest.md]: In C8, raising an error didn't seem to cause much confusion, and it later set the stage for how to use `rescue`.  There are many confusing things about testing, but raise didn't seem to be one of them.  

I'll take it out, but we'll need another introduction later, maybe just with the livecode covering cards to introduce it.
